### PR TITLE
This is my submission for Week3.

### PR DESCRIPTION
**Pull request by duokobia**

### CODE REVIEW

File: 02-express-tutorial/02-http-app.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+const http = require("http");
+const { readFileSync } = require("fs");
+
+// get all files
+const homePage = readFileSync("./navbar-app/index.html");
+const homeStyles = readFileSync("./navbar-app/styles.css");
+const homeImage = readFileSync("./navbar-app/logo.svg");
+const homeLogic = readFileSync("./navbar-app/browser-app.js");
```
Comment (taylor-knapp) [inline 02-express-tutorial/02-http-app.js:8]: I don't see the `navbar-app` directory in your repository -- did it get removed? That makes this file fail to execute (though all the code inside of it is correct).

File: 02-express-tutorial/04-express-app.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+// This is where all static resources should go.
+// Using this method, express takes care of the status codes, content type, file path, etc automatically for us.
+app.use(express.static("./public"));
+
+// For the root
+app.get("/", (req, res) => {
+    // There is a res.sendFile method that comes with ExpressJS. This method requires an absolute path.
+    // So we have to use the "path" module to get this path.
+    // path.resolve or path.join can be used interchangeably in this case. 
+    res.sendFile(path.resolve(__dirname, "./navbar-app/index.html"));
```
Comment (taylor-knapp) [inline 02-express-tutorial/04-express-app.js:23]: Same comment as above about the `navbar-app` directory. This is probably a hidden error b/c the static handler catches the index prior to getting here.

File: 02-express-tutorial/04-express-app.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    }
+);
+// All bad request, use this resource:
+app.all("*", () => {
+    res.status(404).send("resource not found");
+    }
+);
```
Comment (taylor-knapp) [inline 02-express-tutorial/04-express-app.js:30]: In this callback (the second argument passed to `app.all`), you need to define the `req` and `res` parameters so they can be used in the function body. Right now this throws an error.

File: 02-express-tutorial/07-params-query.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+    if (search) {
+        sortedProducts = sortedProducts.filter((product) => {
+            //  We will get errors in the server console if we don't explicitly include "return" 
+            //  in our "if" statements, except it is the last statement in the "if" block. 
+            //  This is because JS will try to send two responses for the same request. 
+            //  It doesnt termnate the previous statement if "return" isn't used. 
+            //  Hence, JS try to send both the current response and the next response at the same time.
+            //  Meanwhile, in JS, you can only have one response per request. 
+            return product.name.startsWith(search);
```
Comment (taylor-knapp) [inline 02-express-tutorial/07-params-query.js:78]: To clarify what's happening here, you are returning a boolean (true/false) if the product in the list starts with the search term. That's separate from what eventually gets returned as the response to the request. (This return statement is actually _within_ the `.filter` method, which isolates it from returning out of the `if` statement.)

File: 02-express-tutorial/practiceExpress.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+const path = require("path");           // Import path module.
+
+const app = express();                  // Invoke Express method.
+
+// This allows Express to take care of the status codes, content type, file path, etc automatically for me.
+app.use(express.static("./new-public"));
+
+app.get("/", (req, res) => {
+    res.sendFile(path.resolve(__dirname, "./new-public/index.html"));
+});
```
Comment (taylor-knapp) [inline 02-express-tutorial/practiceExpress.js:11]: Reminder: this is unnecessary if the static directory has an `index.html` file.

**Review by taylor-knapp — APPROVED** [review]
Great work, and as always, excellent comments! 📜 🖊️ 💯 

A few comments about a missing directory and ensuring you define req/res in your middleware callbacks. You're really getting the concepts!

**Comment by taylor-knapp** [comment]
In regards to your question
> Do I create separate blocks of codes for all nested url or is there a way to handle them as a group using ExpressJS framework? 
> Example: /example, /example/items, /example/items/categories, etc.

I think we'll cover more complex usage patterns in a few more lessons. There are certainly patterns to extract logic into functions that are invoked within route handlers, but you pretty much do always need a specified route + the logic to execute for it for any route a user will use.

File: 02-express-tutorial/02-http-app.js (hunk 1 above)
Comment (duokobia) [reply 02-express-tutorial/02-http-app.js:8]: Yes please.

Thanks for catching this.

I will point them to the directory where the resources are.

File: 02-express-tutorial/07-params-query.js (hunk 4 above)
Comment (duokobia) [reply 02-express-tutorial/07-params-query.js:78]: Thanks for this explanation.

Appreciate!

File: 02-express-tutorial/practiceExpress.js (hunk 5 above)
Comment (duokobia) [reply 02-express-tutorial/practiceExpress.js:11]: Noted Taylor.

I have commented that code block out to keep my code DRY.

Thank you.

**Comment by duokobia** [comment]
Thank you for taking your time to review my codes, helping me understand each concept learnt and catching bugs.

I sincerely appreciate Taylor.

Thank you so much.

**Pull request merged**